File: app/src/main/java/net/twisterrob/android/view/color/swatches/pixel/drawer/StepBitmapDrawer.java
package net.twisterrob.android.view.color.swatches.pixel.drawer;

import net.twisterrob.android.view.color.swatches.pixel.color.PixelColor;

public class StepBitmapDrawer extends BitmapDrawer {
    public StepBitmapDrawer(int[] bitmap, int w, int h, PixelColor pixel) {
        super(bitmap, w, h, pixel);
    }

    @Override
    protected void fillPixels() {
        final int w = this.w;
        final int h = this.h;
        final int iterations = 3;
        final int iterationStep = 1; // must be relative prime to iterations (e.g.: 5,3)
        for (int stepY = 0; stepY < iterations; ++stepY) {
            for (int stepX = 0; stepX < iterations; ++stepX) {
                final int offsetX = (stepX * iterationStep) % iterations;
                final int offsetY = (stepY * iterationStep) % iterations;
                for (int y = offsetY; y < h; y += iterations) {
                    for (int x = offsetX; x < w; x += iterations) {
                        bitmap[y * w + x] = pixel.getPixelColorAt(x, y);
                    }
                }
                reportProgress();
            }
        }
    }
}

File: feature/base/src/main/kotlin/net/twisterrob/android/view/color/swatches/pixel/drawer/StepBitmapDrawer.kt
package net.twisterrob.android.view.color.swatches.pixel.drawer

import androidx.annotation.ColorInt
import androidx.annotation.IntRange
import net.twisterrob.android.view.color.swatches.pixel.color.PixelColor

class StepBitmapDrawer(
	@ColorInt
	bitmap: IntArray,
	@IntRange(from = 0)
	w: Int,
	@IntRange(from = 0)
	h: Int,
	pixel: PixelColor
) : BitmapDrawer(bitmap, w, h, pixel) {

	override fun fillPixels() {
		val w = this.w
		val h = this.h
		val iterations = 3
		val iterationStep = 1 // must be relative prime to iterations (e.g.: 5,3)
		for (stepY in 0 until iterations) {
			for (stepX in 0 until iterations) {
				val offsetX = (stepX * iterationStep) % iterations
				val offsetY = (stepY * iterationStep) % iterations
				// for (y in offsetY until h step iterations)?
				var y = offsetY
				while (y < h) {
					// for (x in offsetX until w step iterations)?
					var x = offsetX
					while (x < w) {
						bitmap[y * w + x] = pixel.getPixelColorAt(x, y)
						x += iterations
					}
					y += iterations
				}
				reportProgress()
			}
		}
	}
}
